### PR TITLE
Add support for offline speculative decoding model PTQ

### PR DESCRIPTION
**Pull request by yeyu-nvidia**
## What does this PR do?

**Type of change:**
new feature

**Overview:** 
This PR enables loading in a ModelOpt pretrained offline speculative decoding model (e.g., EAGLE3) and performs PTQ on it and export.

## Usage
Follow the speculative_decoding examples to train an offline speculative decoding model first.
Then follow the command below to quantize and export it:

```bash
python hf_ptq.py --pyt_ckpt_path <dir_of_offline_specdec_model> --specdec_offline_dataset <dir_of_dataset>
```

## Before your PR is "*Ready for review*"

- **Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/Model-Optimizer/blob/main/CONTRIBUTING.md)** and your commits are signed.
- **Is this change backward compatible?**: Yes/No 
- **Did you write any new necessary tests?**: Yes/No
- **Did you add or update any necessary documentation?**: Yes/No
- **Did you update [Changelog](https://github.com/NVIDIA/Model-Optimizer/blob/main/CHANGELOG.rst)?**: Yes/No 

## Summary by CodeRabbit

* **New Features**
  * Offline speculative decoding workflow: support loading a local dataset for calibration, generation, and export; new CLI option to specify the offline dataset.

* **Improvements**
  * Export and quantization paths now accept and propagate offline speculative-decoding inputs.
  * Offline data loading honors a sample-size limit and enforces safe batch sizing for calibration.

* **Bug Fixes**
  * Better handling of model/config mismatches and varied batch types in offline flows.

### CODE REVIEW

File: examples/llm_ptq/hf_ptq.py [hunk 1]
```diff
@@ -444,28 +451,34 @@ def load_model(args: argparse.Namespace):
...
-            : len(args.dataset)
-        ]
+            if extracted_lm is not None:
+                language_model = extracted_lm
+                model_type = extracted_model_type
+
         tokenizer = get_tokenizer(args.pyt_ckpt_path, trust_remote_code=args.trust_remote_code)
 
         default_padding_side = tokenizer.padding_side
         default_pad_token = tokenizer.pad_token
```
Comment (ChenhanYu) [inline examples/llm_ptq/hf_ptq.py:537]: Typo: `devlazy_preprocessice=True` — this looks like a garbled field name (maybe `lazy_preprocess`?). Since this is an `argparse.Namespace`, the wrong name won't error — it'll be silently ignored, potentially causing incorrect behavior downstream. Please verify and fix.

File: examples/llm_ptq/hf_ptq.py [hunk 2]
```diff
@@ -444,28 +451,34 @@ def load_model(args: argparse.Namespace):
...
+            # We only quantize the language model for VLMs other than the type supported above.
+            extracted_lm, extracted_model_type = extract_and_prepare_language_model_from_vl(
+                full_model
             )
-        # Adjust calib_size to match dataset length by extending or truncating as needed
-        args.calib_size = (args.calib_size + [args.calib_size[-1]] * len(args.dataset))[
-            : len(args.dataset)
-        ]
+            if extracted_lm is not None:
+                language_model = extracted_lm
```
Comment (ChenhanYu) [inline examples/llm_ptq/hf_ptq.py:531]: `sys.path.append` to import from a sibling example directory is fragile — breaks if `__file__` is unset or the relative path changes. Consider making `eagle_utils` a proper importable module, or at minimum wrap in try/except with a helpful error message.

File: examples/llm_ptq/hf_ptq.py [hunk 3]
```diff
@@ -590,6 +603,7 @@ def export_quantized(
     tokenizer: PreTrainedTokenizerBase | None,
     default_padding_side,
     default_pad_token,
+    offline_specdec_input: dict | None = None,
 ):
     with torch.inference_mode():
```
Comment (ChenhanYu) [inline examples/llm_ptq/hf_ptq.py:675]: `next(iter(calib_dataloader), None)` consumes batch 0 as a dummy input for the export forward pass. This works because `DataLoader` creates fresh iterators, but the intent isn't obvious. A comment explaining why this is needed (offline specdec models require their specific input format for the dummy forward) would help.

File: modelopt/torch/utils/dataset_utils.py [hunk 4]
```diff
@@ -519,8 +519,11 @@ def _process_batch(batch_data, infer_method, max_working_batch_size=None):
     Returns:
         The maximum batch size that worked successfully
     """
-    assert all(torch.is_tensor(data) or data is None for data in batch_data.values()), (
```
Comment (ChenhanYu) [inline modelopt/torch/utils/dataset_utils.py:522]: Special-casing `base_model_outputs` by key name in a generic batch processing utility is a code smell — ties this utility to a specific feature. Consider either: (a) adding a parameter to skip validation for certain keys, or (b) normalizing `base_model_outputs` to tensors before it reaches this function.

File: examples/speculative_decoding/main.py [hunk 5]
```diff
@@ -67,7 +67,6 @@ class DataArguments:
     data_path: str = field(
         metadata={"help": "Path to the training data."},
     )
-    eval_data_path: str = field(default=None, metadata={"help": "Path to the evaluation data."})
```
Comment (ChenhanYu) [inline examples/speculative_decoding/main.py]: `eval_data_path` is removed from `DataArguments`. If any existing training scripts or configs reference this field, they'll break. Is this intentional cleanup or accidental?

File: examples/speculative_decoding/eagle_utils.py [hunk 6]
```diff
@@ -158,6 +158,8 @@ def make_eagle_supervised_data_module(
         if not dumped_files:
             raise ValueError(f"No .pt files found in {data_args.offline_data_path}")
```
Comment (ChenhanYu) [inline examples/speculative_decoding/eagle_utils.py:86]: `sample_size > 0` means `sample_size=0` produces an empty list (`[:0]`), and `sample_size=-1` (the default in `main.py`) slices off the last element. The sentinel values are inconsistent. Consider:
```python
if data_args.sample_size > 0:
    dumped_files = dumped_files[:data_args.sample_size]
```
should guard against `-1` explicitly, e.g.:
```python
if data_args.sample_size not in (-1, 0) and data_args.sample_size > 0:
```
or just `if data_args.sample_size > 0:` with documentation that `<= 0` means all. Either way, please add unit tests for the edge cases.

File: examples/llm_ptq/hf_ptq.py (hunk 1 above)
Comment (yeyu-nvidia) [reply examples/llm_ptq/hf_ptq.py:537]: Fixed in 4586572d — corrected to `lazy_preprocess=True`.

File: examples/llm_ptq/hf_ptq.py (hunk 2 above)
Comment (yeyu-nvidia) [reply examples/llm_ptq/hf_ptq.py:531]: Fixed in 4586572d — replaced `sys.path.append` with `importlib.util.spec_from_file_location` to load `eagle_utils` directly by path without polluting `sys.path`. Since this is an examples directory (not a package), making it a proper module isn't appropriate.

File: examples/llm_ptq/hf_ptq.py (hunk 3 above)
Comment (yeyu-nvidia) [reply examples/llm_ptq/hf_ptq.py:675]: Fixed in 4586572d — added a comment explaining that offline speculative decoding models require a real data sample as the dummy input for the export forward pass, since their input format (`base_model_outputs`, etc.) differs from standard models.

File: modelopt/torch/utils/dataset_utils.py (hunk 4 above)
Comment (yeyu-nvidia) [reply modelopt/torch/utils/dataset_utils.py:522]: Fixed in 4586572d — refactored `_process_batch`, `_forward_loop`, and `create_forward_loop` to accept an `allowed_non_tensor_keys: set | None` parameter instead of hardcoding `'base_model_outputs'`. The offline dataset call site now passes `allowed_non_tensor_keys={"base_model_outputs"}` explicitly.

File: examples/speculative_decoding/main.py (hunk 5 above)
Comment (yeyu-nvidia) [reply examples/speculative_decoding/main.py]: `eval_data_path` was intentionally removed — it was unused in the offline training path and the new `offline_data_path` field serves the equivalent purpose for offline data.

File: examples/speculative_decoding/eagle_utils.py (hunk 6 above)
Comment (yeyu-nvidia) [reply examples/speculative_decoding/eagle_utils.py:86]: Fixed in 4586572d — added a comment clarifying that non-positive values (including `-1` and `0`) all use the full dataset. The `-1` convention is documented in the `DataArguments.sample_size` field help text.

**Comment by yeyu-nvidia** [comment]
Addressing the testing feedback from the review (a15a804b):

**Unit tests added:**
- `tests/unit/torch/utils/test_dataset_utils.py`: 3 new tests for `_process_batch` with `allowed_non_tensor_keys` — verifies non-tensor values are accepted under the allowlist, other keys are still validated, and the default (no allowlist) rejects non-tensors.
- `tests/unit/torch/speculative/plugins/test_hf_speculative_offline.py`: 5 tests for `sample_size` truncation edge cases (positive N, -1, 0, N > dataset len, empty dir) and 2 tests for `offline_specdec_input` propagation through `export_speculative_decoding`.

**Comment by yeyu-nvidia** [comment]
**CI test added** (217777cb):

`tests/examples/speculative_decoding/test_eagle_offline_ptq.py` covers the full three-stage pipeline in a single CI job:
1. `test_collect_hidden_states` — runs `compute_hidden_states_hf.py` with 2 conversations and verifies `.pt` files are produced with the correct `input_ids`/`hidden_states` keys.
2. `test_offline_eagle_training` — runs `launch_train.sh --offline-data` with a tiny EAGLE config and verifies the output checkpoint directory is created.
3. `test_offline_ptq` — runs `hf_ptq.py` with `--specdec_offline_dataset` and verifies `model.safetensors` + `config.json` are present with the expected EAGLE weight keys.

The test uses minimal data (2 conversations, `--training_seq_len 64`, `--calib 2`) to keep CI runtime short.

File: examples/llm_ptq/hf_ptq.py [hunk 7]
```diff
@@ -1069,6 +1118,14 @@ def parse_args() -> argparse.Namespace:
...
         default=None,
     )
+    parser.add_argument(
+        "--specdec_offline_dataset",
+        help=(
+            "If set, the model is a speculative decoding model,"
+            "which uses offline dataset for calibration. "
+        ),
+        default=None,
+    )
```
[the remark this answers is not in the document]
Comment (yeyu-nvidia) [reply examples/llm_ptq/hf_ptq.py:1206]: Fixed in 04f83daf — added upfront validation in `parse_args()` to reject `--specdec_offline_dataset` when combined with `--sparsity_fmt != dense`, and in `__main__` to reject comma-separated `--calib` values.

File: examples/llm_ptq/hf_ptq.py
Comment (ChenhanYu) [inline examples/llm_ptq/hf_ptq.py:1]: Will it be cleaner if we separate the the speculative decoding ptq to examples/speculative_decoding?

**Comment by ChenhanYu** [comment]
Also, we need to add an Qwen3-8B EAGLE PTQ example to `tools/launcher/examples` to complete this PR

File: examples/llm_ptq/hf_ptq.py
Comment (yeyu-nvidia) [reply examples/llm_ptq/hf_ptq.py:1]: Depends. Our Online PTQ will use hf_ptq so I would prefer we leave all ptq code together

Comment (yeyu-nvidia) [reply examples/llm_ptq/hf_ptq.py:1]: Adding to the above — the offline PTQ code in `hf_ptq.py` is ~50 lines that reuse the existing PTQ infrastructure (model loading, calibration loop, export). Separating it to `examples/speculative_decoding/` would require duplicating all of that shared code, and as mentioned, the upcoming online PTQ path will also live in `hf_ptq.py`. Keeping it together avoids divergence.

**Comment by yeyu-nvidia** [comment]
Added in the latest push — `tools/launcher/examples/Qwen/Qwen3-8B/hf_offline_eagle3_ptq.yaml` is a 5-step pipeline (query → dump hidden states → train EAGLE3 → PTQ with `--specdec_offline_dataset` → VLLM benchmark). Also added the wrapper script at `tools/launcher/common/eagle3/hf_ptq.sh`.

File: modelopt/torch/export/unified_export_hf.py [hunk 8]
```diff
@@ -378,6 +380,18 @@ def llm_dummy_forward():
...
+            target_device = next(model.parameters()).device
+
+            def _to_device(value):
+                if isinstance(value, torch.Tensor):
+                    return value.to(target_device)
+                if isinstance(value, dict):
+                    return {k: _to_device(v) for k, v in value.items()}
+                return value
+
+            model(**_to_device(offline_specdec_input))
```
Comment (h-guo18) [inline modelopt/torch/export/unified_export_hf.py]: Is `offline_specdec_input` necessary here? Can we just use dummy inputs

File: examples/llm_ptq/hf_ptq.py [hunk 9]
```diff
@@ -896,16 +925,47 @@ def quantize_main(
 
     print(f"Use calib batch_size {args.batch_size}")
 
-    calib_dataloader, first_text_speech_dataset = make_calib_dataloader(
-        args, language_model, processor, tokenizer, device, model_type
-    )
+    if args.specdec_offline_dataset is not None:
+        _eagle_utils_path = os.path.join(
+            os.path.dirname(__file__), "../speculative_decoding/eagle_utils.py"
```
Comment (h-guo18) [inline examples/llm_ptq/hf_ptq.py]: How about moving OfflineSupervisedDataset and EagleOfflineDataCollator to some importable module like `modelopt/torch/speculative/eagle/utils.py`? I think they are critical enough to be put there

File: examples/speculative_decoding/eagle_utils.py [hunk 10]
```diff
@@ -124,6 +125,15 @@ def __call__(self, features: list[dict[str, Any]]) -> dict[str, Any]:
...
+        if self.device is not None:
+            batch = {
+                k: v.to(self.device)
+                if isinstance(v, torch.Tensor)
+                else {kk: vv.to(self.device) for kk, vv in v.items()}
+                if isinstance(v, dict)
+                else v
+                for k, v in batch.items()
+            }
         return batch
```
Comment (h-guo18) [inline examples/speculative_decoding/eagle_utils.py]: I think moving data between devices in data collator may mess with dataloader's multithreading/prefetching logic. Conventionally data moving is taken care by the trainer, rather than data collator.

This new argument "device" is confusing for eagle training usage, where it should be left None.

I suggest move this data-moving logic from data collator to make_calib_dataloader() instead.

File: examples/llm_ptq/hf_ptq.py [hunk 11]
```diff
@@ -896,16 +925,47 @@ def quantize_main(
 
     print(f"Use calib batch_size {args.batch_size}")
 
-    calib_dataloader, first_text_speech_dataset = make_calib_dataloader(
-        args, language_model, processor, tokenizer, device, model_type
-    )
+    if args.specdec_offline_dataset is not None:
```
Comment (h-guo18) [inline examples/llm_ptq/hf_ptq.py]: Since `make_calib_dataloader` already handles a few similar branching logic and have all the args we need, shall we move this if-branch inside `make_calib_dataloader` for a cleaner interface?

File: examples/llm_ptq/hf_ptq.py (hunk 9 above)
Comment (h-guo18) [reply examples/llm_ptq/hf_ptq.py]: Otherwise this import seems a little hacky to me to put into ptq entrance

File: examples/llm_ptq/hf_ptq.py [hunk 12]
```diff
@@ -988,17 +1048,18 @@ def quantize_main(
...
-        full_model,
-        model_type,
-        tokenizer,
-        processor,
-        preview_input_ids,
-        generated_ids_before_ptq,
-        is_nemotron_vl_model,
-        first_text_speech_dataset,
-    )
+    if args.specdec_offline_dataset is None:
```
Comment (h-guo18) [inline examples/llm_ptq/hf_ptq.py]: How about moving this branch inside  `post_quantize` as a early-exiting branch?
Reason:
1. There are already similar branching logic like [this](https://github.com/NVIDIA/Model-Optimizer/blob/main/examples/llm_ptq/hf_ptq.py#L812) inside post_quantize(). This if-branch does similar branching.
2. It leaves a cleaner interface in ptq main entrance.

File: examples/llm_ptq/hf_ptq.py [hunk 13]
```diff
@@ -896,16 +925,47 @@ def quantize_main(
...
+            args, language_model, processor, tokenizer, device, model_type
+        )
 
     # Detect if this is a Nemotron VL model using architecture-based detection
     is_nemotron_vl_model = is_nemotron_vl(full_model)
 
-    preview_input_ids, generated_ids_before_ptq = pre_quantize(
-        args, full_model, model_type, tokenizer, calib_dataloader, is_nemotron_vl_model
-    )
+    if args.specdec_offline_dataset is None:
```
Comment (h-guo18) [inline examples/llm_ptq/hf_ptq.py]: similarly, shall we hide this branch inside pre_quantize?

File: examples/llm_ptq/hf_ptq.py [hunk 14]
```diff
@@ -346,7 +348,13 @@ def forward_step(model, batch):
 def load_model(args: argparse.Namespace):
     # If low memory mode is enabled, we compress the model while loading the HF checkpoint.
     calibration_only = False
-    if not args.low_memory_mode:
+    if args.specdec_offline_dataset is not None:
```
Comment (h-guo18) [inline examples/llm_ptq/hf_ptq.py]: Shall we do something if both specdec_offline_dataset and low_memory_mode is set?

File: examples/llm_ptq/hf_ptq.py
Comment (ChenhanYu) [reply examples/llm_ptq/hf_ptq.py:1]: OK

**Comment by ChenhanYu** [comment]
OK, we need to add the ptq launcher test to nmm-sandbox. Approve to unblock.

**Review by ChenhanYu — APPROVED** [review]

File: modelopt/torch/export/unified_export_hf.py (hunk 8 above)
Comment (yeyu-nvidia) [reply modelopt/torch/export/unified_export_hf.py]: We could construct dummy inputs from the model config (hidden_size, etc.), but that would couple the export pipeline to EAGLE's internal data format — if the format changes, the dummy construction breaks silently. Using real `offline_specdec_input` is more robust since it always matches what the model actually expects. The trade-off isn't worth it IMO.

File: examples/llm_ptq/hf_ptq.py (hunk 9 above)
Comment (yeyu-nvidia) [reply examples/llm_ptq/hf_ptq.py]: Done in ac825fb — moved `OfflineSupervisedDataset` and `EagleOfflineDataCollator` to `modelopt/torch/speculative/eagle/utils.py`. The old location in `eagle_utils.py` re-imports for backward compat.

File: examples/speculative_decoding/eagle_utils.py (hunk 10 above)
Comment (yeyu-nvidia) [reply examples/speculative_decoding/eagle_utils.py]: Good point — fixed in ac825fb. Removed `device` parameter from `EagleOfflineDataCollator`. Device transfer is now handled by a `_DeviceDataLoader` wrapper in `make_calib_dataloader()`, keeping the collator clean for both training and PTQ usage.

File: examples/llm_ptq/hf_ptq.py (hunk 11 above)
Comment (yeyu-nvidia) [reply examples/llm_ptq/hf_ptq.py]: Done in ac825fb — the offline dataset/dataloader creation is now inside `make_calib_dataloader()` alongside the existing VLM/whisper/text branches.

File: examples/llm_ptq/hf_ptq.py (hunk 9 above)
Comment (yeyu-nvidia) [reply examples/llm_ptq/hf_ptq.py]: Fixed in ac825fb — the `importlib` hack is gone. Now imports directly from `modelopt.torch.speculative.eagle.utils`.

File: examples/llm_ptq/hf_ptq.py (hunk 12 above)
Comment (yeyu-nvidia) [reply examples/llm_ptq/hf_ptq.py]: Done in ac825fb — moved the offline specdec export into `post_quantize()` as an early-exit branch, following the existing pattern. Main entrance is cleaner now.

File: examples/llm_ptq/hf_ptq.py (hunk 14 above)
Comment (yeyu-nvidia) [reply examples/llm_ptq/hf_ptq.py]: Good catch — `low_memory_mode` compresses during loading which doesn't apply to the offline path. Added validation in `parse_args()` in commit 04f83da to reject unsupported combinations. Let me add this one too.

File: examples/llm_ptq/hf_ptq.py (hunk 13 above)
Comment (yeyu-nvidia) [reply examples/llm_ptq/hf_ptq.py]: Done in 345be77 — moved the guard inside `pre_quantize()` as an early return. Also added `--low_memory_mode` validation in `parse_args()`.

File: modelopt/torch/export/unified_export_hf.py (hunk 8 above)
Comment (h-guo18) [reply modelopt/torch/export/unified_export_hf.py]: How about we implement a `get_dummy_offline_inputs()` for class HFEagleModel and all incoming SD algorithms like dflash? Then we keep the eagle-specific logics locolized, without introducing new arguments all the way from export_quantized() -> export_speculative_decoding() -> exporter -> _export_transformers_checkpoint() -> requantize_resmooth_fused_llm_layers()

Comment (h-guo18) [reply modelopt/torch/export/unified_export_hf.py]: then we simply `model(**_to_device(model.get_dummy_inputs()))` here, regardless of what algorithm it is

Comment (yeyu-nvidia) [reply modelopt/torch/export/unified_export_hf.py]: Great idea — implemented in 0b7b960. Added `get_dummy_inputs()` to `HFEagleModel` that constructs the right input format (with `base_model_outputs` for offline models, plain `input_ids` otherwise). The export pipeline now just does `model(**model.get_dummy_inputs())` when `get_dummy_inputs()` is available. Removed `offline_specdec_input` from the entire chain: `export_quantized` → `export_speculative_decoding` → `exporter.export` → `_export_transformers_checkpoint` → `requantize_resmooth_fused_llm_layers`. This is extensible for future SD algorithms like dflash — they just implement their own `get_dummy_inputs()`.

File: modelopt/torch/speculative/eagle/utils.py [hunk 15]
```diff
@@ -70,3 +76,83 @@ def expand_mask(mask: torch.Tensor, dtype: torch.dtype, tgt_len: int | None = No
...
+    ):
+        """Initialize with a list of .pt file paths."""
+        super().__init__()
+        self.dumped_files = dumped_files
+
+    def __len__(self):
+        return len(self.dumped_files)
+
+    def __getitem__(self, i) -> dict[str, torch.Tensor]:
+        offline_data = torch.load(self.dumped_files[i])
```
Comment (shengliangxu) [inline modelopt/torch/speculative/eagle/utils.py]: weights_only=True

File: modelopt/torch/export/unified_export_hf.py [hunk 16]
```diff
@@ -752,7 +755,9 @@ def _export_transformers_checkpoint(
 
     # Resmooth and requantize fused layers
     # TODO: Handle mixed precision
-    requantize_resmooth_fused_llm_layers(model)
+    requantize_resmooth_fused_llm_layers(
+        model
+    )
```
Comment (meenchen) [inline modelopt/torch/export/unified_export_hf.py]: This change seems unnecessary.

File: modelopt/torch/speculative/eagle/utils.py [hunk 17]
```diff
@@ -70,3 +76,83 @@ def expand_mask(mask: torch.Tensor, dtype: torch.dtype, tgt_len: int | None = No
     inverted_mask = 1.0 - expanded_mask
 
     return inverted_mask.masked_fill(inverted_mask.to(torch.bool), torch.finfo(dtype).min)
+
+
+class OfflineSupervisedDataset(Dataset):
+    """Offline dataset for supervised fine-tuning.
+
+    This dataset loads data on-the-fly from pre-processed .pt data files.
+
```
Comment (meenchen) [inline modelopt/torch/speculative/eagle/utils.py]: Can you add more context about the dataset? like how to generate it and the expected format?

File: examples/llm_ptq/hf_ptq.py [hunk 18]
```diff
@@ -346,7 +397,13 @@ def forward_step(model, batch):
...
+        model_kwargs = {"trust_remote_code": args.trust_remote_code}
+        if args.attn_implementation is not None:
+            model_kwargs["attn_implementation"] = args.attn_implementation
+        full_model = AutoModelForCausalLM.from_pretrained(args.pyt_ckpt_path, **model_kwargs)
+        full_model = full_model.to(args.device)
```
Comment (meenchen) [inline examples/llm_ptq/hf_ptq.py]: Could you share why `get_model` doesn't work for `args.specdec_offline_dataset`?

File: modelopt/torch/speculative/eagle/utils.py (hunk 15 above)
Comment (yeyu-nvidia) [reply modelopt/torch/speculative/eagle/utils.py]: Done in 59674e2 — added `weights_only=True` to `torch.load` in `OfflineSupervisedDataset.__getitem__`.

File: modelopt/torch/export/unified_export_hf.py (hunk 16 above)
Comment (yeyu-nvidia) [reply modelopt/torch/export/unified_export_hf.py]: Good catch — reverted in d4aada4.

File: modelopt/torch/speculative/eagle/utils.py (hunk 17 above)
Comment (yeyu-nvidia) [reply modelopt/torch/speculative/eagle/utils.py]: Added more context in d4aada4. The docstring now documents the expected `.pt` file format (keys: `input_ids`, `hidden_states`, `aux_hidden_states`, `base_model_input_embeds`) and how to generate them (`main.py --mode dump_offline_data`).

File: examples/llm_ptq/hf_ptq.py (hunk 18 above)
Comment (yeyu-nvidia) [reply examples/llm_ptq/hf_ptq.py]: \`get_model\` sets \`device_map=\"auto\"\` which distributes the model across GPUs. For offline specdec PTQ, the eagle module gets attached later via \`mtq.convert()\` and needs all base model parameters on a single device for the dummy forward pass and quantization calibration. Using \`get_model\`'s automatic device mapping could conflict with this.

We could refactor \`get_model\` to accept a device_map override, but the simpler direct loading is more explicit for this use case. Happy to refactor if you feel strongly.

Comment (meenchen) [reply examples/llm_ptq/hf_ptq.py]: `needs all base model parameters on a single device for the dummy forward pass` This seems to be a tight constraint. Do we have a plan for large models?

Comment (yeyu-nvidia) [reply examples/llm_ptq/hf_ptq.py]: Good question. The single-device constraint only applies during the `requantize_resmooth` export forward pass (one dummy forward through the EAGLE module), not during calibration. Calibration already uses `device_map="auto"` via `get_model` for the base model's forward passes.

For the offline PTQ path specifically, we skip the base model forward entirely (hidden states are pre-computed), so the model loaded here is only used for:
1. `mtq.convert()` to attach quantization nodes to the EAGLE module
2. Calibration forward passes through the EAGLE module only (tiny — ~200M params)
3. The dummy forward in `requantize_resmooth`

Since the EAGLE module itself is always small regardless of base model size, the single-device placement is not a practical bottleneck — even for 70B+ base models, the EAGLE head is ~200M-400M params.

If we ever need to support very large base models in a non-offline PTQ path, we can extend `get_dummy_inputs()` to handle multi-device placement. But for the offline path, this is a non-issue.

Comment (yeyu-nvidia) [reply examples/llm_ptq/hf_ptq.py]: You're right — my earlier reply was wrong about the constraint being a non-issue. The offline path loads the **full base model** (not just the EAGLE head), so a 70B model at ~140GB BF16 won't fit on a single GPU.

Fixed in e07af925: switched the offline path to `device_map="auto"` (matching `get_model()`'s behavior). HF accelerate's dispatch hooks handle cross-device tensor movement during the forward pass, so both calibration and the `requantize_resmooth` dummy forward work correctly with multi-GPU sharding.

Comment (yeyu-nvidia) [reply examples/llm_ptq/hf_ptq.py]: Actually, simplified further in 766a9e1 — just reuse `get_model()` directly instead of a separate loading path. The two flags (`--specdec_offline_dataset` and `--low_memory_mode`) are mutually exclusive (validated at argparse level), so the conditions merge cleanly into `if args.specdec_offline_dataset is not None or not args.low_memory_mode`. No more duplicated model loading logic.
